### PR TITLE
Sof input device demo

### PR DESCRIPTION
**Pull request by jsarha**

### CODE REVIEW

**Comment by jsarha** [comment]
This draft PR is not intended to be merged. It is just a demo of how to deliver input events from SOF FW to Linux input device framework. The PR depends on this SOF FW PR: https://github.com/thesofproject/sof/pull/5847

**Comment by plbossart** [comment]
@jsarha is this PR missing the code for the input client proper?

Also this could really added as a generic client providing notifications over IPC. There will be other cases where we need to provide 'triggers' or information about 'acoustic events' back to userspace.

**Comment by ranj063** [comment]
> This draft PR is not intended to be merged. It is just a demo of how to deliver input events from SOF FW to Linux input device framework. The PR depends on this SOF FW PR: [thesofproject/sof#5847](https://github.com/thesofproject/sof/pull/5847)

@jsarha what is an "input device"? Do you have an example of what the niput events from the FW look like and what they might be used for?

**Comment by jsarha** [comment]
> > This draft PR is not intended to be merged. It is just a demo of how to deliver input events from SOF FW to Linux input device framework. The PR depends on this SOF FW PR: [thesofproject/sof#5847](https://github.com/thesofproject/sof/pull/5847)
> 
> @jsarha what is an "input device"? Do you have an example of what the niput events from the FW look like and what they might be used for?

For instance keyboard and mouse are input devices. I guess "human input device" would be more accurate term. This would enable SOF FW to generate input events for Linux UI.

**Comment by jsarha** [comment]
> input client proper?

**Comment by jsarha** [comment]
> @jsarha is this PR missing the code for the input client proper?
> 

I am not sure if I follow the question, but the idea was to take the events up to /dev/input/event? and leave the rest to the user space.

File: include/sound/sof/header.h [hunk 1]
```diff
@@ -53,6 +53,7 @@
 #define SOF_IPC_GLB_TEST_MSG			SOF_GLB_TYPE(0xBU)
 #define SOF_IPC_GLB_PROBE			SOF_GLB_TYPE(0xCU)
 #define SOF_IPC_GLB_DEBUG			SOF_GLB_TYPE(0xDU)
+#define SOF_IPC_GLB_INPUT_EVENT			SOF_GLB_TYPE(0xFU)
```
Comment (ranj063) [inline include/sound/sof/header.h:56]: wouldn;t this be a major ABI change

**Comment by jsarha** [comment]
Forgot "git add sound/soc/sof/sof-input-device.c". Now the input device driver code is there (Thanks to Ranjani).

File: sound/soc/sof/sof-input-device.c [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+#include <sound/sof/input-event.h>
+#include <linux/input.h>
+
+struct sof_input_device {
+	struct input_dev *input_dev;
+	struct device *dev;
+};
+
+static
+void snd_sof_input_device_event(struct sof_client_dev *cdev, void *msg_buf)
```
Comment (ranj063) [inline sound/soc/sof/sof-input-device.c]: i think we should drop the snd prefix

File: sound/soc/sof/sof-input-device.c [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+	struct device *dev = &auxdev->dev;
+	struct sof_input_device *sid;
+	int ret;
+
+	if (sof_client_get_ipc_type(cdev) != SOF_IPC)
+		return -ENOTSUPP;
+
+	sid = devm_kzalloc(dev, sizeof(*sid), GFP_KERNEL);
+	if (!sid) {
+		dev_err(dev, "%s:%d: Not enough memory\n", __func__,
```
Comment (ranj063) [inline sound/soc/sof/sof-input-device.c]: i dont think we print error messages for ENOMEM anywhere else

File: sound/soc/sof/sof-input-device.c [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+		return ret;
+	}
+
+	sid->dev = dev;
+	cdev->data = sid;
+
+	ret = sof_client_register_ipc_rx_handler(cdev, SOF_IPC_GLB_INPUT_EVENT,
+						 snd_sof_input_device_event);
+	if (ret)
+		goto register_ipc_rx_err;
```
Comment (ranj063) [inline sound/soc/sof/sof-input-device.c]: since it is only used in 1 place, might be just better to unregister and return here instead of the goto

**Comment by plbossart** [comment]
Yes @jsarha this really looks like an HID driver, so in that case it might be useful to have the HID report sent over IPC so that it can be provided directly to HID drivers.

This is how SoundWire planned to deal with e.g. button press in codec driver, you have an HID template and report that are used to pass the information in generic ways to upper levels of the stack.

File: include/sound/sof/header.h (hunk 1 above)
Comment (jsarha) [reply include/sound/sof/header.h:56]: Yes, it is. This is just a proof of concept. I think the final version should be implemented for IPC4 in the first place, but I do not understand that well enough yet (working on it ATM).

**Comment by jsarha** [comment]
Addressed @ranj063 's review comments despite it is unclear when this work will continue.

**Comment by jsarha** [comment]
> Yes @jsarha this really looks like an HID driver, so in that case it might be useful to have the HID report sent over IPC so that it can be provided directly to HID drivers.
> 
> This is how SoundWire planned to deal with e.g. button press in codec driver, you have an HID template and report that are used to pass the information in generic ways to upper levels of the stack.

Ok. Interesting indeed. That could offer at least a good starting point for the message needed. For the demo I just wanted have minimal implementation of all the steps needed, without too much concentrating on what information should be transferred, as it would anyway be application specific and almost impossible to anticipate in early stage.

**Comment by plbossart** [comment]
Closing since there's no activity and no plan to merge.

**Pull request closed without merging**